feathr_project/feathr/feature_derivations.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+from typing import List, Optional
+
+from copy import copy, deepcopy
+from jinja2 import Template
+from typing import Union
+from feathr.dtype import FeatureType, ValueType
+from feathr.feature import Feature
+from feathr.typed_key import DUMMY_KEY, TypedKey
+from feathr.transformation import ExpressionTransformation, RowTransformation
+
+class DerivedFeature:
+    """A derived feature is a feature defined on top of other features, rather than external data source.
+    Attributes:
+        name: derived feature name
+        feature_type: type of derived feature
+        key: join key of the derived feature
+        input_features: features that this derived features depends on
+        transform: transformation that produces the derived feature value, based on the input_features
+        feature_alias: Rename the derived feature to `feature_alias`. Default to None.
+        key_alias: Rename the key of derived feature to `key`. Default to None.
+    """
+
+    def __init__(self,
+                name: str,
+                feature_type: FeatureType,
+                input_features: Union[DerivedFeature, Feature, List[Union[DerivedFeature, Feature]]],
+                transform: Union[str, RowTransformation],
+                key: Optional[Union[TypedKey, List[TypedKey]]] = [DUMMY_KEY],
+                feature_alias:Optional[str] = None,
+                key_alias: Optional[List[str]] = None) -> None:
+        self.name = name
+        self.feature_type = feature_type
+        self.key = key if isinstance(key, List) else [key]
+        self.input_features = input_features if isinstance(input_features, List) else [input_features]
```

Comment:
Can we do some validation such as that the keys used by the `input_features` are in the `key` list/set?

This way, we can prevent user from shooting themselves in the foot